### ISSUE TITLE
Bug with swab, unexpected result - fix

### ISSUE THREAD

**Issue by michelerenzullo** [emscripten-core/emscripten#15578]
[swab.c](https://github.com/emscripten-core/emscripten/blob/f0c4614a279ff57ce331ba3bcb60b076ac171603/system/lib/libc/musl/src/string/swab.c)

I found this bug using a library in my webassembly project, after 2 days I found the source of the divergence.

This function swap the bytes, the problem is if compiled with any c++ compiler it produces the correct value, if compiled with Emscripten it produces the wrong value because during the swap it "doesn't remember" the original value before it has been replaced, so he copy 2 times the same value. I simply add a "tmp" value that stores the original value to be swapped.

```
void swab_fixed(const char* src, char * dest, size_t n)
{   uint8_t tmp;
	for (; n > 1; n -= 2) {
		tmp = dest[0];
		dest[0] = src[1];
		dest[1] = tmp;
		dest += 2;
		src += 2;
	}
}
```

I enclosed a simple test you can verify the issue, compile both with emscripten and with your standard c++ compiler, you will notice the difference.
```
em++ test.cpp -o test.js
node test.js
```

test.cpp:
```
#include <string>
#include <unistd.h>

void swab_fixed(const char* src, char * dest, size_t n)
{   uint8_t tmp;
	for (; n > 1; n -= 2) {
		tmp = dest[0];
		dest[0] = src[1];
		dest[1] = tmp;
		dest += 2;
		src += 2;
	}
}

bool check(uint16_t test) {

	if (test != 344) {
		printf("output is %d, not 344, Emscripten error!\n", test);
		return false;
	}
	else printf("output is %d, OK!\n", test);
	return true;
}

int main()
{
  unsigned count = 1;
	uint16_t * test = new uint16_t[count];
	test[0] = 22529;
	
	swab((char *) test, (char *) test, count * 2);
	if (!check(test[0])) {
		printf("Retrying with fixed swab...\n"); 
		test[0] = 22529; swab_fixed((char *) test, (char *) test, count * 2); 
		check(test[0]);
	}

	delete[] test;
	return 0;
}
```

**Comment by kripken** [comment]
I'm not an expert on these standards, but [the docs](https://man7.org/linux/man-pages/man3/swab.3.html) defines it like this:

```c
void swab(const void *restrict from, void *restrict to, ssize_t n);
```

The musl libc code does the same. The `restrict` annotation indicates that the system can assume the inputs do *not* overlap as they do in your testcase. That is, I think you are hitting undefined behavior. It is possible that other libcs do not use the undefined behavior to optimize, but it looks like musl does?

**Comment by michelerenzullo** [comment]
Yes, is true. But it looks like that all other libcs doesn't optimize with UB. Maybe I should report to the Lib repo I'm working on to replace or create a copy of dst.

**Comment by michelerenzullo** [comment]
Any update about?
